projects/skps/codecs2/skps_heatmap.py

```diff
@@ -0,0 +1,166 @@
+        - heatmap size: [W, H]
+
+    Encoded:
+
+        - heatmaps (np.ndarray): The generated heatmap in shape (K, H, W)
+            where [W, H] is the `heatmap_size`
+        - offside (np.ndarray): The generated offside in x andy direction
+            in shape (2K, H, W) where [W, H] is the `heatmap_size`
+        - keypoint_weights (np.ndarray): The target weights in shape (N, K)
+
+    Args:
+        input_size (tuple): Image size in [w, h]
+        heatmap_size (tuple): Heatmap size in [W, H]
+        sigma (float): The sigma value of the Gaussian heatmap
+    """
+
+    def __init__(self, input_size: Tuple[int, int],
+                 heatmap_size: Tuple[int, int], sigma: float) -> None:
+        super().__init__()
+        self.input_size = input_size
+        self.heatmap_size = heatmap_size
+        self.sigma = sigma
+        self.scale_factor = (np.array(input_size) /
+                             heatmap_size).astype(np.float32)
+
+        self.y_range, self.x_range = np.meshgrid(
+            np.arange(0, self.heatmap_size[1]),
+            np.arange(0, self.heatmap_size[0]),
+            indexing='ij')
+
+    def encode(self,
+               keypoints: np.ndarray,
+               keypoints_visible: Optional[np.ndarray] = None) -> dict:
+        """Encode keypoints into heatmaps. Note that the original keypoint
+        coordinates should be in the input image space.
+
+        Args:
+            keypoints (np.ndarray): Keypoint coordinates in shape (N, K, D)
+            keypoints_visible (np.ndarray): Keypoint visibilities in shape
+                (N, K)
+
+        Returns:
+            dict:
+            - heatmaps (np.ndarray): The generated heatmap in shape
+                (K, H, W) where [W, H] is the `heatmap_size`
+            - offside (np.ndarray): The generated offside in x and y
+                directions in shape (2*K, H, W) where [W, H] is the
+                `heatmap_size`
+            - keypoint_weights (np.ndarray): The target weights in shape
+                (N, K)
+        """
+
+        assert keypoints.shape[0] == 1, (
+            f'{self.__class__.__name__} only support single-instance '
+            'keypoint encoding')
+
+        if keypoints_visible is None:
+            keypoints_visible = np.ones(keypoints.shape[:2], dtype=np.float32)
+
+        heatmaps, keypoint_weights = generate_unbiased_gaussian_heatmaps(
+            heatmap_size=self.heatmap_size,
+            keypoints=keypoints / self.scale_factor,
+            keypoints_visible=keypoints_visible,
+            sigma=self.sigma)
+
+        offside_heatmap = self.generate_offside_heatmap(
+            heatmap_size=self.heatmap_size,
+            keypoints=keypoints / self.scale_factor,
+        )
+
+        encoded = dict(
+            heatmaps=heatmaps,
+            keypoint_weights=keypoint_weights[0],
+            displacements=offside_heatmap)
+
+        return encoded
+
+    def generate_offside_heatmap(
```

Comment:
For 'offside', do you mean 'offset'?

Reply:
'generate_offset_map' seems better. Because it is not actually heatmaps.

Reply:
Yes, it's better, I will check them.
And another question. For gt labels returned, which key is better for this offset_map. I place it with displacements now, is it ok?